packages/world/src/systems/CraftFuelSystem.sol

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { System } from "@latticexyz/world/src/System.sol";
+
+import { ObjectType } from "../codegen/tables/ObjectType.sol";
+
+import { InventorySlot, InventorySlotData } from "../codegen/tables/InventorySlot.sol";
+import { ObjectTypeMetadata } from "../codegen/tables/ObjectTypeMetadata.sol";
+
+import { CRAFT_ENERGY_COST } from "../Constants.sol";
+import { EntityId } from "../EntityId.sol";
+import { ObjectTypeId } from "../ObjectTypeId.sol";
+import { ObjectTypeLib } from "../ObjectTypeLib.sol";
+import { ObjectTypes } from "../ObjectTypes.sol";
+
+import { Vec3 } from "../Vec3.sol";
+import { transferEnergyToPool } from "../utils/EnergyUtils.sol";
+import { InventoryUtils, SlotAmount } from "../utils/InventoryUtils.sol";
+import { CraftNotification, notify } from "../utils/NotifUtils.sol";
+
+contract CraftFuelSystem is System {
+  using ObjectTypeLib for ObjectTypeId;
+
+  function craftFuel(EntityId caller, EntityId powerstone, SlotAmount[] memory inputs) public {
+    caller.activate();
+    require(inputs.length > 0, "Must provide at least one input");
+    require(
+      powerstone.exists() && ObjectType._get(powerstone) == ObjectTypes.Powerstone,
+      "You need a powerstone to craft fuel"
+    );
+    caller.requireConnected(powerstone);
+
+    uint128 totalEnergy = 0;
+    for (uint256 i = 0; i < inputs.length; i++) {
+      ObjectTypeId inputType = InventorySlot._getObjectType(caller, inputs[i].slot);
+      require(inputType.isLog() || inputType.isLeaf(), "Can only use logs or leaves to make fuel");
+      // we convert the mass to energy
+      totalEnergy +=
+        inputs[i].amount * (ObjectTypeMetadata._getEnergy(inputType) + ObjectTypeMetadata._getMass(inputType));
+      InventoryUtils.removeObjectFromSlot(caller, inputType, inputs[i].amount, inputs[i].slot);
+    }
+    uint128 fuelAmount = totalEnergy / ObjectTypeMetadata._getEnergy(ObjectTypes.Fuel);
```

Comment:
I wonder if we should just do fuelAmount == totalEnergy, we could have rounding errors if not. Unless we know that the energy + mass of all logs and leaves is always a multiple of fuel

Reply:
the energy of fuel is set to 1 right now so it should be a multiple

I can remove the division though and save a read

Reply:
oh, but in that case won't fuel be too many to fit in the inventory? oak log has 12500000000000000 mass

Reply:
haha I mean its 1 without scaling, so fuel = 100000000000000 mass

1 oak log will give you 178 (125 mass + 53 energy) fuel

Reply:
ah nice ok  👍

Reply:
hardcoded the energy value with a comment